src/Shared/Dictionary/SmallCapacityDictionary.cs

```diff
@@ -0,0 +1,760 @@
+        }
+
+        /// <inheritdoc />
+        IEnumerator<KeyValuePair<TKey, TValue>> IEnumerable<KeyValuePair<TKey, TValue>>.GetEnumerator()
+        {
+            if (_backup != null)
+            {
+                return _backup.GetEnumerator();
+            }
+
+            return GetEnumerator();
+        }
+
+        /// <inheritdoc />
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            if (_backup != null)
+            {
+                return _backup.GetEnumerator();
+            }
+
+            return GetEnumerator();
+        }
+
+        /// <inheritdoc />
+        bool ICollection<KeyValuePair<TKey, TValue>>.Remove(KeyValuePair<TKey, TValue> item)
+        {
+            if (_backup != null)
+            {
+                return ((ICollection<KeyValuePair<TKey, TValue>>)_backup).Remove(item);
+            }
+
+            if (Count == 0)
+            {
+                return false;
+            }
+
+            var index = FindIndex(item.Key);
+            var array = _arrayStorage;
+            if (index >= 0 && EqualityComparer<TValue>.Default.Equals(array[index].Value, item.Value))
+            {
+                Array.Copy(array, index + 1, array, index, _count - index);
+                _count--;
+                array[_count] = default;
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <inheritdoc />
+        public bool Remove(TKey key)
+        {
+            if (key == null)
+            {
+                ThrowArgumentNullExceptionForKey();
+            }
+
+            if (_backup != null)
+            {
+                return _backup.Remove(key);
+            }
+
+            if (Count == 0)
+            {
+                return false;
+            }
+
+            var index = FindIndex(key);
+            if (index >= 0)
+            {
+                _count--;
+                var array = _arrayStorage;
+                Array.Copy(array, index + 1, array, index, _count - index);
+                array[_count] = default;
+
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Attempts to remove and return the value that has the specified key from the <see cref="SmallCapacityDictionary{TKey, TValue}"/>.
+        /// </summary>
+        /// <param name="key">The key of the element to remove and return.</param>
+        /// <param name="value">When this method returns, contains the object removed from the <see cref="SmallCapacityDictionary{TKey, TValue}"/>, or <c>null</c> if key does not exist.</param>
+        /// <returns>
+        /// <c>true</c> if the object was removed successfully; otherwise, <c>false</c>.
+        /// </returns>
+        public bool Remove(TKey key, out TValue? value)
+        {
+            if (key == null)
+            {
+                ThrowArgumentNullExceptionForKey();
+            }
+
+
+            if (_backup != null)
```

Comment:
Nit: one extra new-line.
```suggestion
            }

            if (_backup != null)
```